StreamSelectLoop: Int overflow for big timer intervals

kaduev13 opened the pull request:
### The problem

As described <a href="https://github.com/reactphp/event-loop/issues/19">here</a>, current `StreamSelectLoop` implementation restricts using time intervals more than `PHP_INT_MAX` microseconds. For 32-bit systems it's only 2148 seconds.

### Suggested solution

* Change `StreamSelectLoop` `$timeout` type from `int` (in microseconds) to `float` (in seconds);
* Replace `usleep` with `time_nanosleep`, cause `usleep` takes `int` argument in microseconds, and we can't keep more than `PHP_INT_MAX` microseconds. But `time_nanosleep` takes 2 arguments: `int` amount of seconds and `int` amount of nanoseconds, that allows `StreamSelectLoop` to handle `PHP_INT_MAX` seconds intervals;
* Extract `stream_select` and `time_nanosleep` methods calls from `StreamSelectLoop::streamSelect(...)` to test timeout values.

jsor commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,117 @@ private function waitForStreamActivity($timeout)
+    {
+        if ($seconds > 0 || $nanoseconds > 0) {
+            time_nanosleep($seconds, $nanoseconds);
+        }
```
I'd like to see explicit handling of `$seconds`or `$nanoseconds`being `null` here.

jsor commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,117 @@ private function waitForStreamActivity($timeout)
+            return PHP_INT_MAX;
+        }
+
+        return $time === null ? null : intval(floor($time));
```
Minor nitpicking: for consistency with the other 2 methods, i'd like to an early return for `$time === null`:

```php
if ($time === null) {
    return null;
}

// ...

return intval(floor($time));
```

jsor commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,124 @@ private function waitForStreamActivity($timeout)
+    private static function getMicroseconds($time)
+    {
+        if ($time === null) {
+            return null;
```
Should probably return `0`. `stream_select` accepts `null` only for the `$seconds` parameter.

jsor commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,124 @@ private function waitForStreamActivity($timeout)
+    private static function getNanoseconds($time)
+    {
+        if ($time === null) {
+            return null;
```
Should probably return `0`, see above.

jsor commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,124 @@ private function waitForStreamActivity($timeout)
+     * @param array $write
+     * @param array $except
+     * @param int|null $seconds
+     * @param int|null $microseconds
```
Should only accept `int`, see also above.

kaduev13 commented:
@jsor, thank you for the review. I think that previous implementation was not explicit, cause `$timeout` checking for `null` value was duplicated in 4 places (`getSeconds`, `getMicroseconds`, `getNanoseconds` and `sleep`). I changed this behavior in c78624d: I removed all unnecessary `null` checks and moved them inside `streamSelect` method. Also, I fixed the things you pointed ;)

jsor commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,112 @@ private function waitForStreamActivity($timeout)
     {
+        $seconds = $timeout === null ? null : self::getSeconds($timeout);
+        $microseconds = $timeout === null ? 0 : self::getMicroseconds($timeout);
+        $nanoseconds = $timeout === null ? 0 : self::getNanoseconds($timeout);
```
Calculation of `$microseconds` and `$nanoseconds` can probably be moved into the `if`s so they are only calculated when needed.

kaduev13 replied:
@jsor, done in bb81383 😃

jsor commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,112 @@ private function waitForStreamActivity($timeout)
+         * (int)(float)PHP_INT_MAX == PHP_INT_MAX => false
+         * (int)(float)PHP_INT_MAX == PHP_INT_MIN => true
+         */
+        if ($time == PHP_INT_MAX) {
```
Strict comparision (`===`)?

kaduev13 replied:
@jsor, `$time` is float, so strict comparison will always return `false`.

```php
(float) PHP_INT_MAX === PHP_INT_MAX; //is false
(float) PHP_INT_MAX == PHP_INT_MAX; //is true
```

jsor replied:
👍 (maybe add a comment that loose comparision is intentional)

kaduev13 replied:
@jsor I added more description about this tricky  `if` in 3c8a5d6 😄

jsor commented:
ping @clue

kaduev13 commented:
@clue, can you please describe, why did you mark this PR as BC Break? Imo, only internal representation of time changed and no public interfaces modified...

clue commented on src/StreamSelectLoop.php:
```diff
@@ -220,27 +213,117 @@ private function waitForStreamActivity($timeout)
      * @param array        &$read   An array of read streams to select upon.
      * @param array        &$write  An array of write streams to select upon.
-     * @param integer|null $timeout Activity timeout in microseconds, or null to wait forever.
+     * @param float|null   $timeout Activity timeout in seconds, or null to wait forever.
```
This is a rather subtle BC break as consumers *could* possibly rely on this `protected` API. Not saying this is an issue with your proposed change, it's more that we ~screwed up~ could have done better with defining our external API.
